man/splitpatch.1
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "splitpatch 1"
.TH splitpatch 1 "2014-06-19" "splitpatch" "2014-06-19"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
splitpatch \- split a patch up into files or hunks
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&  splitpatch [options]
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Divide a patch or diff file into pieces. The split can made by file
or by hunk basis. This makes is possible to separate changes that
might not be desirable, or assemble a patch into more coherent set
of changes.
.PP
The hunk option makes it possible to compare similar patches on a
hunk-by-hunk basis using a tool like \fIinterdiff\fR\|(1) from the patchutils package.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-H, \-\-hunk, \-\-hunks\fR" 4
.IX Item "-H, --hunk, --hunks"
Split patch by hunks instead of the default: by file.
.IP "\fB\-h, \-\-help\fR" 4
.IX Item "-h, --help"
Display help and exit
.IP "\fB\-V, \-\-version\fR" 4
.IX Item "-V, --version"
Output version information and exit.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Have you ever been working on code, gone off on a side tangent and
then realized you don't have a patch for the original issue you
set out to address?  When you run \f(CW\*(C`diff \-u\*(C'\fR or some other diff command,
you discover that you have made multiple sets of changes; some
you want to submit, others you don't. What you really want to do is
break those changes apart and apply only the ones you want. This is
where splitpatch comes into play, breaking up the patch by source file:
.PP
.Vb 1
\&    splitpatch changes.patch
.Ve
.PP
Or to split the patch file into individual hunks which can later be
applied using the \fIpatch\fR\|(1) command:
.PP
.Vb 1
\&    splitpatch \-\-hunks changes.patch
.Ve
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
None.
.SH "FILES"
.IX Header "FILES"
The \fB\-\-hunk\fR option writes sequentially numbered files in the current
directory named for the source file to be patched followed by the 
sequence number: \f(CW\*(C`*\-NNN.patch\*(C'\fR.
.PP
Otherwise, the split patch files are named by the source file being
patched.
.SH "STANDARDS"
.IX Header "STANDARDS"
None.
.SH "AVAILABILITY"
.IX Header "AVAILABILITY"
https://github.com/benjsc/splitpatch
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIdehtmldiff\fR\|(1)
\&\fIeditdiff\fR\|(1)
\&\fIfilterdiff\fR\|(1)
\&\fIfixcvsdiff\fR\|(1)
\&\fIflipdiff\fR\|(1)
\&\fIgrepdiff\fR\|(1)
\&\fIinterdiff\fR\|(1)
\&\fIlsdiff\fR\|(1)
\&\fIpatch\fR\|(1)
\&\fIrecountdiff\fR\|(1)
\&\fIrediff\fR\|(1)
\&\fIsplitdiff\fR\|(1)
\&\fIunwrapdiff\fR\|(1)
\&\fIwiggle\fR\|(1)
.SH "AUTHORS"
.IX Header "AUTHORS"
Program was written by Peter Hutterer <peter@cs.unisa.edu.au>.
.PP
This manual page was written by Jari Aalto <jari.aalto@cante.net>.
Released under license \s-1GNU GPL\s0 version 2 or (at your option) any later
version. For more information about the license, visit
<http://www.gnu.org/copyleft/gpl.html>.


splitpatch.rb
#!/usr/bin/env ruby
#
#   Copyright
#
#       Copyright (C) 2014 Jari Aalto <jari.aalto@cante.net>
#       Copyright (C) 2007-2014 Peter Hutterer <peter.hutterer@who-t.net>
#       Copyright (C) 2007-2014 Benjamin Close <Benjamin.Close@clearchain.com>
#
#   License
#
#       This program is free software; you can redistribute it and/or modify
#       it under the terms of the GNU General Public License as published by
#       the Free Software Foundation; either version 2 of the License, or
#       (at your option) any later version.
#
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#       GNU General Public License for more details.
#
#       You should have received a copy of the GNU General Public License
#       along with this program. If not, see <http://www.gnu.org/licenses/>.
#
#  Description
#
PROGRAM = "splitpatch"
VERSION = 1.0
LICENSE = "GPL-2+"  # See official acronyms: https://spdx.org/licenses/
AUTHOR  = "Peter Hutterer <peter.hutterer@who-t.net>"

#       Splitpatch is a simple script to split a patch up into
#       multiple patch files. If the --hunks option is provided on the
#       command line, each hunk gets its own patchfile.

class Splitter
    def initialize(file)
       @filename = file
       @fullname = false
    end

    def fullname(opt)
       @fullname = opt
    end

    def validFile?
        return File.exist?(@filename) && File.readable?(@filename)
    end

    def createFile(filename)
        if File.exists?(filename)
            puts "File #{filename} already exists. Renaming patch."
            appendix = 0
            zero = appendix.to_s.rjust(3, '0')
            while File.exists?("#{filename}.#{zero}")
                appendix += 1
                zero = appendix.to_s.rjust(3, '0')
            end
            filename << ".#{zero}"
        end
        return open(filename, "w")
    end

    def getFilename(line)
        tokens = line.split(" ")
        tokens = tokens[1].split(":")
        tokens = tokens[0].split("/")
        if @fullname
            return tokens.join('-')
        else
            return tokens[-1]
        end
    end

    # Split the patchfile by files 
    def splitByFile
        legacy = false
        outfile = nil
        stream = open(@filename)
        until (stream.eof?)
            line = stream.readline

            # we need to create a new file
            if (line =~ /^Index: .*/) == 0
                # patch includes Index lines
                # drop into "legacy mode"
                legacy = true
                if (outfile)
                    outfile.close_write
                end
                filename = getFilename(line)
                filename << ".patch"
                outfile = createFile(filename)
                outfile.write(line)
            elsif (line =~ /--- .*/) == 0 and not legacy
                if (outfile) 
                    outfile.close_write
                end
                #find filename
                filename = getFilename(line)
                filename << ".patch"
                outfile = createFile(filename)
                outfile.write(line)
            else
                if outfile
                    outfile.write(line)
                end
            end
        end
    end

    def splitByHunk
        legacy = false
        outfile = nil
        stream = open(@filename)
        filename = ""
        counter = 0
        header = ""
        until (stream.eof?)
            line = stream.readline

            # we need to create a new file
            if (line =~ /^Index: .*/) == 0
                # patch includes Index lines
                # drop into "legacy mode"
                legacy = true
                filename = getFilename(line)
                header = line
                # remaining 3 lines of header
                for i in 0..2
                    line = stream.readline
                    header << line
                end
                counter = 0
            elsif (line =~ /--- .*/) == 0 and not legacy
                #find filename
                filename = getFilename(line)
                header = line
                # next line is header too
                line = stream.readline
                header << line
                counter = 0
            elsif (line =~ /@@ .* @@/) == 0
                if (outfile) 
                    outfile.close_write
                end

                zero = counter.to_s.rjust(3, '0')
                hunkfilename = "#{filename}.#{zero}.patch"
                outfile = createFile(hunkfilename)
                counter += 1

                outfile.write(header)
                outfile.write(line)
            else
                if outfile
                    outfile.write(line)
                end
            end
        end
    end

end

def help
    puts <<EOF
SYNOPSIS
    #{PROGRAM} [options] FILE.patch

OPTIONS
    -h,--help
    -H,--hunk
    -V,--version

DESCRIPTION

    Split the patch up into files or hunks

    Divide a patch or diff file into pieces. The split can made by file
    or by hunk basis. This makes it possible to separate changes that
    might not be desirable or assemble the patch into a more coherent set
    of changes. See e.g. combinediff(1) from patchutils package.

    Note: only patches in unified format are recognized.
EOF
end

def version
  puts "#{VERSION} #{LICENSE} #{AUTHOR}"
end

########################     MAIN     ########################

if ARGV.length < 1 or ARGV.length > 2
    puts "ERROR: missing argument. See --help."
    exit 1
else
    opt = ARGV[0]
    if /^-h$|--help/.match(opt)
        help
        exit 0
    elsif /^-H$|--hunks?/.match(opt)
        hunk = 1
    elsif /^-V$|--version/.match(opt)
        version
        exit 0
    elsif /^-/.match(opt)
        puts "ERROR: Unknonw option: #{opt}. See --help."
        exit 1
        end
    file = ARGV[-1]
    s = Splitter.new(file)
    if s.validFile?
        if hunk
            s.splitByHunk
        else
            s.splitByFile
        end
    else
        puts "File does not exist or is not readable: #{file}"
    end
end

# End of file
